### ISSUE TITLE
simplify remote plugins

### ISSUE THREAD

**Issue by justinmk** [neovim/go-client#167]
# Problem

The design, and usage, of remote plugins can be simplified, as described in https://github.com/neovim/neovim/issues/27949 

tl;dr:

- remove the concept of "remote plugins"
- instead, any Go module that imports the go-client can call `set_handler('foo', myFunction)` to handle the "foo" request from Nvim
- "remote plugins" become plain old Lua plugins which call RPC methods on the Go module

# Solution

- Apply the above proposal to go-client. 
- Deprecate the old rplugin interface.

# Reference

- Implementation in node-client: https://github.com/neovim/node-client/pull/344
- discussion/samples of current go-client interface: https://github.com/neovim/neovim/discussions/27371

**Comment by zchee** [comment]
@justinmk I don't understand all the new remote plugin behavior yet, but I decided to work on the migration process ASAP.

**Comment by garyburd** [comment]
> remove the concept of "remote plugins"

Delete the [plugin](https://pkg.go.dev/github.com/neovim/go-client@v1.2.1/nvim/plugin) package and any references to that package in the documentation

> instead, any Go module that imports the go-client can call set_handler('foo', myFunction) to handle the "foo" request from Nvim

The package accomplishes this goal as is:

**Step 1** The application [creates a new Nvim client](https://pkg.go.dev/github.com/neovim/go-client/nvim#New) connected to stdio:

```
// Turn off timestamps in output.
log.SetFlags(0) 

// Create a client connected to stdout. Configure the client
// to use the standard log package for logging.
v, err := nvim.New(os.Stdin, os.Stdout, os.Stdout, log.Printf)
if err != nil {
    log.Fatal(err)
}

// Redirect the application's direct use of stdout to stderr.
// Direct writes by the application garble the RPC stream.
os.Stdout = os.Stderr
```

**Step 2** The application registers handlers with the nvim client using the [client's RegisterHandler method](https://pkg.go.dev/github.com/neovim/go-client/nvim#Nvim.RegisterHandler).

**Step 3**  The application runs the the RPC message loop.

```
if err := v.Serve(); err != nil {
    log.Fatal(err)
}
```

**Optional enhancement** Encapsulate the code above in a helper function:

```
package nvim

import (
    "log"
    "os"
)

// RemoteModuleMain implements the main function for a Nvim remote module.
//
// Applications call the RemoteModuleMain function to run the module. The
// function creates an Nvim client, calls the supplied function to register
// handlers with the client and then runs the server loop to handle requests
// from Nvim.
//
// Applications should use the default logger in the standard log package to
// write to Nvim's log.
func RemoteModuleMain(registerHandlers func(v *Nvim) error) {
    log.SetFlags(0) 
    v, err := nvim.New(os.Stdin, os.Stdout, os.Stdout, log.Printf)
    if err != nil {
        log.Fatal(err)
    }
    os.Stdout = os.Stderr
    if err := registerHandlers(v); err != nil {
       log.Fatal(err)
    }
    if err := v.Serve(); err != nil {
        log.Fatal(err)
    }
}

```

Example use:

```
package main

import "github.com/neovim/go-client/nvim"

func clearCurrentLine(v *nvim.Nvim) {
     v.SetCurrentLine([]byte{})
}

func main() {
    nvim.RemoteModuleMain(func(v *Nvim) error {
        v.RegisterHandler("clearCurrentLine", clearCurrentLine)
    })
}

```

**Comment by justinmk** [comment]
@garyburd thanks so much, that is very helpful. So this is basically already supported by go-client.

> **Optional enhancement** Encapsulate the stdio related code in a helper function.

💯

**Comment by garyburd** [comment]
@justinmk Yes, it's already supported.  I propose the following:

- Prepare for plugin deprecation.
  - Move [plugin documentation](https://github.com/neovim/go-client/blob/9b44fee9cf8b370d8b3372565be66a5208c05bc5/README.md#L9-L70) from the top-level README.md  to new file nvim/plugin/README.md.
  - Add deprecation notice to the  [nvim/plugin package comment](https://github.com/neovim/go-client/blob/main/nvim/plugin/main.go#L1) and the new README file.  Include link to neovim/issues/27949 in the notice.
- Add new directory examples/remote containing the following:
    - ./plugin/goexample.lua: remote module Lua code.
    - ./goexample/{main.go, go.mod}: remote module Go code
    - ./README.md: instructions for building and installing the remote module.

  The immediate goal is to post minimal working code using the package as is. Let's wait for progress on neovim/issues/27949 before adding the proposed helper function, writing complete documentation and adding tests.

I have some time for this in the next couple of days. @zchee and @justinmk let me know your thoughts on the proposal.

**Comment by zchee** [comment]
@garyburd cc: @justinmk 
Totally understand and almost agreed to your proposal. Thanks for quickly improvement!

More comment later, because now JST is start working time.

**Comment by justinmk** [comment]
> The immediate goal is to post minimal working code using the package as is.

Sounds good!

> Let's wait for progress on neovim/issues/27949 before adding the proposed helper function, writing complete documentation and adding tests.

27949 won't be a "big bang", we'll have to do it incrementally.

Since go-client already has `RegisterHandler`, if we can provide a full example that shows a Lua plugin starting a go-client process and calling its methods via `vim.rpcrequest`/`vim.rpcnotify`, there isn't really anything blocking us from officially deprecating the go-client rplugin support.

This will be doubly helpful, because it will be an end-to-end demonstration of the logistics for 27949.

Of course, in the future we might add nice-to-have features in Nvim that make it more convenient to spin up a long-lived client and call its RPC methods. But I don't see that as a requirement for this phase.

**Comment by garyburd** [comment]
@justinmk I wanted to wait on 27949 before adding an API in case the work on 27949 exposes additional requirements for the API.  If you think things are baked enough, then I'll proceed ahead with the API and deprecation.   I may not be able to get to it for a couple of weeks. 

Here's a list of public Go plugins:  https://pkg.go.dev/github.com/neovim/go-client/nvim/plugin?tab=importedby

**Comment by justinmk** [comment]
> in case the work on 27949 exposes additional requirements for the API

We may weave-in more nice-to-have ideas later, but that will just upgrade the UX rather than be a hard requirement.

Currently, the (informal) requirements for a remote plugin are:

1. [responds to "poll" ](https://github.com/neovim/neovim/blob/2f4792943aa92223fadd472f20449cd13707ff7a/runtime/autoload/provider.vim#L13) (I guess this is more for "providers" than rplugins)
    - seems potentially worth keeping as a basic "liveness check".
3. returns a ["spec"](https://github.com/neovim/neovim/blob/2f4792943aa92223fadd472f20449cd13707ff7a/runtime/autoload/remote/host.vim#L84)
    - can probably be dropped. remote plugins will be coupled with their Lua setup code, which inherently defines the "spec".
